### DISCUSSION TITLE
HTTP sender API 'Bad Request" error and cURL command issue

### DISCUSSION THREAD

**Discussion by parkinsonr** [nextgenhealthcare/connect#5728] [Q&A] [answered]
I'm attepmting to POST to a token API to request an access token using a client ID and secret. WHen running a cURL command from the commandline of the workstation everything works fine, but I cannot get it to work in Mirth using the HTTP Sender destination, or using a javacript destination and having mirth execute the cURL command using the command line function.

When using the HTTP Sender destination I recieve a "bad request" repsonse from the server

![image](https://user-images.githubusercontent.com/57762761/225981136-66b9c001-d988-4602-8346-7c3b6af37ca5.png)

When using the executeRuntimeCommand in Mirth to execute a cURL command, I receive an "invalid_grant_type" error. (the REMOVED placholders is where i took out sensitive data, i also put hafl the line on a new line so i could fit it in the screenshot but its actually all on 1 line in my channel.
![image](https://user-images.githubusercontent.com/57762761/225981879-37affbd4-6584-4dd9-8b43-b264016ae6c2.png)

When I execute the cURL command manaully in commandline it works fine and I receive the token back:

curl -X POST -H "Content-Type: application/x-www-form-urlencoded" https://REMOVED/api/oauth/token -d "client_id=REMOVED&client_secret=REMOVED&grant_type=client_credentials"

Anyone know what I'm missing? I'd prefer to use the HTTP Sender destination for a cleaner setup but was just trying the command line function to get something to work.

**Comment by pacmano1** [comment]
Post both requests to a truly dumb http listener like npm's http-echo-server.   I't trivial from there to sort out the difference between the mirth post and the curl post.

If you are on Linux and have a recent version of node, just run  `npx http-echo-server` which will start a dumb listener on port 3000.   Post to that port and compare the differences between curl and mirth.

**Reply by pacmano1** [reply]
I would expect your parameters section in Mirth to be filled out, it is not currently in your screenshot.

**Reply by parkinsonr** [reply]
Thank you, i did have to add to the query parameters for this content type which I didnt realize.

**Comment by jonbartels** [comment, marked as answer]
https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_HTTP_Sender_dest_con_connect_ug.html section R.

You should remove the header then put your parameters as the form values you want to send. When you set Content Type to `application/x-www-form-urlencoded` MC will automatically handle the form POST correctly.

**Reply by parkinsonr** [reply]
THANK YOU! I didnt put them in the query parameters as I didn't think that was right but after adding those, and also finding out I only had "x-www-form-urlencoded" in the mirth Content-type field, isntead of "application/x-www-form-urlencoded", it's now working. When I copy/pasted into that field I must not have grabbed the application/ part, I just thought it was cut off becuase the box isn't very large.

**Reply by pacmano1** [reply]
That box being to small in several places drives me up the wall!

**Comment by rjkroll** [comment]
Hi,

Not sure if this would help, but we have to post files to CMS via an API.   We had trouble with the HTTP sender, so we just invoked CURL through a Javascript Writer destination via the following, which uses a JWT token retrieved from another channel.   The source is a File Reader where we get the fileName.   filePath is the full path to the file, returned before this step as wel.

```
function runCurl(fileName,filePath,token){
	var path = filePath;
	if(!filePath.endsWith("\\")){ path +="\\";}
	var cmd = 'C:\\Program Files\\Git\\mingw64\\bin\\curl.exe curl -X POST -H "x-file-name:'+ fileName + 
	'" -H "Authorization: ' + token + '" -F "file=@' +path + fileName + '" https://api.endpoint/v1/upload';
	logger.info("command:" + cmd);
	var result = executeRuntimeCommand(cmd);	
	logger.info('Exit value: ' + result.exitValue);
	logger.info('Output: ' + result.stdout);
	if (result.stderr != '') {
		logger.error('Errors: ' + result.stderr);
	}
	channelMap.put('curlOutput',result.stdout);

	channelMap.put('curlExit',result.exitValue);	
	channelMap.put('curlStdErr',result.stderr);	
	if(parseInt(result.exitValue,10) !== 0){return false;/*"Error in curl command"*/}		
	router.routeMessageByChannelId('166476d0-aacf-4e06-9553-2da83d900b37', result.stdout);
	return true;
}

var curlResult = runCurl(sourceMap.get('originalFilename'),sourceMap.get('fileDirectory'), globalMap.get('JWTtoken'));
if(!curlResult){return ERROR;}
return SENT;
```

**Reply by parkinsonr** [reply]
thank you for the info. i ended up getting MC to handle it so no need for the cURL command but thanks anyways.
